addons/binding/org.openhab.binding.airquality/src/main/java/org/openhab/binding/airquality/handler/AirQualityHandler.java

```diff
@@ -0,0 +1,321 @@
+ * @author Łukasz Dywicki
+ */
+public class AirQualityHandler extends BaseThingHandler {
+
+    private Logger logger = LoggerFactory.getLogger(AirQualityHandler.class);
+
+    private static final String URL = "http://api.waqi.info/feed/%QUERY%/?token=%APIKEY%";
+
+    private static final int DEFAULT_REFRESH_PERIOD = 30;
+
+    private ScheduledFuture<?> refreshJob;
+
+    AirQualityJsonResponse aqiResponse;
+
+    private Gson gson;
+
+    public AirQualityHandler(Thing thing) {
+        super(thing);
+        gson = new Gson();
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing Air Quality handler.");
+        super.initialize();
+
+        AirQualityConfiguration config = getConfigAs(AirQualityConfiguration.class);
+        logger.debug("config apikey = (omitted from logging)");
+        logger.debug("config location = {}", config.getLocation());
+        logger.debug("config stationId = {}", config.getStationId());
+        logger.debug("config refresh = {}", config.getRefresh());
+
+        boolean validConfig = true;
+
+        if (StringUtils.trimToNull(config.getApiKey()) == null) {
+            logger.error("Parameter 'apikey' is mandatory and must be configured, disabling thing '{}'",
+                    getThing().getUID());
+            validConfig = false;
+        }
+        if (StringUtils.trimToNull(config.getLocation()) == null && config.getStationId() == null) {
+            logger.error(
+                    "Parameter 'location' or 'stationId' is mandatory and must be configured, disabling thing '{}'",
+                    getThing().getUID());
+            validConfig = false;
+        }
+        if (config.getRefresh() != null && config.getRefresh() < 5) {
+            logger.error("Parameter 'refresh' must be at least 5 minutes, disabling thing '{}'", getThing().getUID());
+            validConfig = false;
+        }
+
+        if (validConfig) {
+            startAutomaticRefresh();
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
+        }
+    }
+
+    /**
+     * Start the job refreshing the Air Quality data
+     */
+    private void startAutomaticRefresh() {
+        if (refreshJob == null || refreshJob.isCancelled()) {
+            Runnable runnable = new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        // Request new air quality data to the aqicn.org service
+                        aqiResponse = updateAirQualityData();
+
+                        // Update all channels from the updated AQI data
+                        for (Channel channel : getThing().getChannels()) {
+                            updateChannel(channel.getUID().getId(), aqiResponse);
+                        }
+                    } catch (Exception e) {
+                        logger.error("Exception occurred during execution: {}", e.getMessage());
```

Comment:
Check against [the guidelines](http://docs.openhab.org/developers/development/guidelines.html#e-logging): Is it a bug if this code is reached? Then also provide the `e` as a last parameter, so that the stack trace is included. Or can this also e.g. happen if the HTTP request times out? Then rather set the Thing status instead.